### 재용님이 보내주신 전처리 코드 이용해서 증감 형식으로 된 것들을 하나의 셀로 합쳤습니다.
### 데이터 어떻게 구성되어있는지 쭉 살펴보시면 됩니다! 

In [2]:
import pandas as pd


# CSV 파일 불러오기
file_path = r"C:\Users\김영우\기계학습_프로젝트\data\filled\filled_merged_all.csv"
df = pd.read_csv(file_path, encoding="utf-8")

df

순위                영화명         개봉일       매출액  매출액(점유율)    매출액증감  \
0       7               1917  2020-02-19   9214640     0.063  1942660   
1      50              돈 조반니         NaN         0     0.000        0   
2      27             불멸의 연인  1995-01-28    324000     0.002  -106000   
3      48  극장판 미니특공대: 공룡왕 디노  2020-02-06     15000     0.000    15000   
4      45    아르코라이브 뮤지컬 안테모사  2020-03-25     48000     0.000    48000   
...    ..                ...         ...       ...       ...      ...   
73805   3                 파과  2025-04-30  36139800     0.151  3198900   
73806   2              썬더볼츠＊  2025-04-30  41381180     0.173  4379500   
73807   1                 야당  2025-04-16  61031100     0.256  4201800   
73808  19                광란자         NaN    579000     0.002   579000   
73809  29      5시부터 7시까지 클레오         NaN    110000     0.000   -46000   

       매출액증감율(전일대비)       누적매출액   관객수  관객수증감(전일대비)  ...  상영횟수  대표국적        국적  \
0             0.267  2932328200  1445          547  ...   108    미국        미국   
1             1.000           0     0            0  ...     3    영국        영국   
2            -0.247     7284000    50          -11  ...     5    미국        미국   
3             1.000    92172600     2            2  ...     1    한국        한국   
4             1.000     4533000     4            4  ...     2    한국        한국   
...             ...         ...   ...          ...  ...   ...   ...       ...   
73805         0.097  1250043260  3774          280  ...   300    한국        한국   
73806         0.118  2810394940  4061          556  ...   333    미국        미국   
73807         0.074  6798977980  6346          479  ...   645    한국        한국   
73808         1.000      579000    83           83  ...     1    서독     서독,미국   
73809        -0.295    15085000    11           -2  ...     1   프랑스  프랑스,이탈리아   

                                 제작사                    배급사         등급  \
0                                NaN              (주)스마일이엔티   15세이상관람가   
1                                NaN                    NaN        NaN   
2      아이콘엔터테인먼트인터내셔널,머제스틱 필름즈 인터내셔널              (주)엔케이컨텐츠  고등학생이상관람가   
3                    (주)에스에이엠지엔터테인먼트              (주)리틀빅픽쳐스      전체관람가   
4                          한국문화예술위원회               (주)아이디어랩      전체관람가   
...                              ...                    ...        ...   
73805                         (주)수필름    (주)넥스트엔터테인먼트월드(NEW)   15세이상관람가   
73806                            NaN     월트디즈니컴퍼니코리아 유한책임회사   12세이상관람가   
73807                    (주)하이브미디어코프  메가박스중앙(주) 플러스엠 엔터테인먼트    청소년관람불가   
73808                            NaN                    NaN        NaN   
73809                            NaN            (주)영화사 안다미로        NaN   

                장르        감독  \
0           드라마,전쟁     샘 멘데스   
1               공연    카스퍼 홀텐   
2              드라마    버나드 로즈   
3            애니메이션       이영준   
4           공연,뮤지컬   이정민,오미선   
...            ...       ...   
73805  액션,드라마,미스터리       민규동   
73806      액션,어드벤처  제이크 슈레이어   
73807        범죄,액션       황병국   
73808   드라마,스릴러,범죄  윌리엄 프리드킨   
73809          드라마   아녜스 바르다   

                                                      배우         조회일  
0      조지 맥케이,딘-찰스 채프먼,콜린 퍼스,베네딕트 컴버배치,마크 스트롱,앤드류 스캇,...  2020-04-29  
1                                                    NaN  2020-04-29  
2                               게리 올드만,예로엔 크라베,이사벨라 로셀리니  2020-04-29  
3      신용우,엄상현,양정화,이소영,전태열,남도형,소연,홍진욱,정영웅,김은아,정영웅,이민규...  2020-04-29  
4                강지혜,장예원,고은영,김찬종,김대호,문경초,장두환,김담현,정지은,전성혜  2020-04-29  
...                                                  ...         ...  
73805            이혜영,김성철,연우진,김무열,신시아,옥자연,김강우,최무성,양주미,양주미  2025-05-14  
73806  플로렌스 퓨,세바스찬 스탠,와이어트 러셀,올가 쿠릴렌코,크리스 바우어,데이빗 하버,...  2025-05-14  
73807            강하늘,유해진,박해준,류경수,채원빈,김민규,이건영,송하선,우아름,오충근  2025-05-14  
73808                                                NaN  2025-05-14  
73809                                 코린느 마르샹,앙트완느 부르세이에  2025-05-14  

[73810 rows x 23 columns]

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 73810 entries, 0 to 73809
Data columns (total 23 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   순위            73810 non-null  int64  
 1   영화명           73810 non-null  object 
 2   개봉일           60766 non-null  object 
 3   매출액           73810 non-null  int64  
 4   매출액(점유율)      73810 non-null  float64
 5   매출액증감         73810 non-null  int64  
 6   매출액증감율(전일대비)  73810 non-null  float64
 7   누적매출액         73810 non-null  int64  
 8   관객수           73810 non-null  int64  
 9   관객수증감(전일대비)   73810 non-null  int64  
 10  관객수증감율(전일대비)  73810 non-null  float64
 11  누적관객수         73810 non-null  int64  
 12  스크린수          73810 non-null  int64  
 13  상영횟수          73810 non-null  int64  
 14  대표국적          73568 non-null  object 
 15  국적            73568 non-null  object 
 16  제작사           23974 non-null  object 
 17  배급사           62652 non-null  object 
 18  등급            61994 non-nu

In [4]:
# 결측치가 있는 컬럼과 결측치 개수만 출력
missing_cols = df.isnull().sum()
missing_cols = missing_cols[missing_cols > 0]  # 결측치가 1개 이상인 컬럼만 필터링

print(missing_cols)

개봉일     13044
대표국적      242
국적        242
제작사     49836
배급사     11158
등급      11816
장르        298
감독       3446
배우      15962
dtype: int64


In [5]:
# 1. 중복 제거 (영화 제목 기준으로)
df_unique = df.drop_duplicates(subset="영화명")

# 2. 결측치가 있는 컬럼과 그 개수 출력
missing_cols = df_unique.isnull().sum()
missing_cols = missing_cols[missing_cols > 0]

print(missing_cols)

개봉일     4011
대표국적      47
국적        47
제작사     5233
배급사     3838
등급      3888
장르        62
감독      1844
배우      3819
dtype: int64


In [6]:
counts = df['영화명'].value_counts()
counts

너의 이름은.        856
날씨의 아이         791
더 퍼스트 슬램덩크     677
라라랜드           489
극장판 주술회전 0     464
              ... 
밀접접촉             1
뺑반               1
수호지-무림호걸 이운      1
라이징 썬            1
광란자              1
Name: 영화명, Length: 6650, dtype: int64

In [7]:
counts = df['영화명'].value_counts()
count_threshold = (counts >= 3).sum()
print(count_threshold)

2860


In [8]:
# 2. 4번 이상 등장하는 영화명만 추출
frequent_titles = counts[counts >= 4].index

# 3. 해당 영화명들만 포함하는 데이터프레임 정의
df_filtered = df[df['영화명'].isin(frequent_titles)]

# 4. 결측치가 있는 컬럼과 결측치 개수 확인
missing_cols = df_filtered.isnull().sum()
missing_cols = missing_cols[missing_cols > 0]

missing_cols

개봉일      7746
대표국적      194
국적        194
제작사     44262
배급사      6002
등급       6599
장르        236
감독        924
배우      11215
dtype: int64

In [9]:
# 1. '영화명' 등장 횟수 계산
counts = df['영화명'].value_counts()

# 2. 4번 이상 등장하는 영화명만 추출
frequent_titles = counts[counts >= 4].index

# 3. 해당 영화명들만 포함하는 데이터프레임 정의
df_filtered = df[df['영화명'].isin(frequent_titles)]

# 4. 영화명 기준으로 중복 제거 (대표 1개씩만)
df_unique = df_filtered.drop_duplicates(subset='영화명')

# 5. 결측치가 있는 컬럼과 결측치 개수 확인
missing_cols = df_unique.isnull().sum()
missing_cols = missing_cols[missing_cols > 0]

print(missing_cols)

개봉일      787
대표국적      19
국적        19
제작사     1750
배급사      687
등급       727
장르        26
감독       100
배우       636
dtype: int64


In [10]:
import pandas as pd

def merge_duplicate_movies_include_all(df):
    movie_col = '영화명'
    count_cols = ['매출액증감', '관객수증감(전일대비)']
    keep_cols = ['개봉일', '대표국적', '국적', '배급사', '등급', '장르', '감독']
    list_cols = ['매출액', '관객수', '스크린수', '상영횟수', '조회일']
    mean_cols = ['매출액(점유율)']
    last_cols = ['누적매출액', '누적관객수']

    # 1. 존재하는 컬럼만 필터링
    count_cols = [col for col in count_cols if col in df.columns]
    keep_cols = [col for col in keep_cols if col in df.columns]
    list_cols = [col for col in list_cols if col in df.columns]
    mean_cols = [col for col in mean_cols if col in df.columns]
    last_cols = [col for col in last_cols if col in df.columns]

    # 2. 조회일을 datetime으로 변환 (정렬을 위해)
    if '조회일' in df.columns:
        df['조회일'] = pd.to_datetime(df['조회일'], errors='coerce')

    # 3. 중복 영화명 추출
    counts = df[movie_col].value_counts()
    dup_titles = counts[counts > 1].index

    # 4. 중복 영화만 필터링 & 조회일 기준 최신순 정렬
    df_dup = df[df[movie_col].isin(dup_titles)].sort_values(by=[movie_col, '조회일'], ascending=[True, False])

    # 5. 병합 방식 정의
    agg_dict = {col: lambda x: list(x) for col in count_cols + list_cols}
    agg_dict.update({col: 'first' for col in keep_cols})
    agg_dict.update({col: 'mean' for col in mean_cols})
    agg_dict.update({col: 'last' for col in last_cols})

    # 6. groupby로 병합
    df_merged = df_dup.groupby(movie_col).agg(agg_dict).reset_index()
    df_merged['중복횟수'] = df_dup[movie_col].value_counts()[df_merged[movie_col]].values

    # 7. 중복 아닌 영화는 그대로 유지
    df_unique = df[~df[movie_col].isin(dup_titles)].copy()
    df_unique['중복횟수'] = 1

    # 8. 병합 + 단일영화 결합
    df_final = pd.concat([df_merged, df_unique], ignore_index=True)

    return df_final

In [11]:
result_df = merge_duplicate_movies_include_all(df_filtered)
result_df.head()

c:\users\김영우\appdata\local\programs\python\python37\lib\site-packages\ipykernel_launcher.py:20: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy


영화명  \
0                     #살아있다   
1                     #해시태그   
2  (여자)아이들 월드투어 [아이돌] 인 시네마   
3             007 노 타임 투 다이   
4                       1+1   

                                               매출액증감  \
0  [2890000, -1266000, 3707000, 2004000, 39000, -...   
1  [65300, 63500, 27700, 27500, -16000, -120200, ...   
2  [600000, -3875000, 225000, 2675000, -2075000, ...   
3  [2919000, 1120500, 891500, 3145500, 122000, -7...   
4                       [-20000, 24000, 28000, 4000]   

                                         관객수증감(전일대비)  \
0  [289, -253, 742, 400, 6, -5, -6, 4, 2, -5, -10...   
1                           [7, 6, 1, 2, -1, -9, 21]   
2  [24, -155, 9, 107, -83, -38, 180, -16, -177, 1...   
3  [583, 224, 178, 687, 19, -79, 57, -16, -15, -1...   
4                                      [-5, 6, 7, 1]   

                                                 매출액  \
0  [2890000, 4445000, 5711000, 2004000, 46000, 70...   
1  [322500, 257200, 193700, 166000, 138500, 15450...   
2  [1775000, 1175000, 5050000, 4825000, 2150000, ...   
3  [2919000, 5762000, 4641500, 3750000, 604500, 4...   
4                         [4000, 24000, 32000, 4000]   

                                                 관객수  \
0  [289, 889, 1142, 400, 7, 1, 6, 12, 8, 6, 11, 1...   
1                       [27, 20, 14, 13, 11, 12, 21]   
2  [71, 47, 202, 193, 86, 169, 207, 27, 43, 220, ...   
3  [583, 1152, 928, 750, 63, 44, 123, 66, 82, 97,...   
4                                       [1, 6, 8, 1]   

                                                스크린수  \
0  [2, 5, 8, 3, 1, 1, 1, 1, 2, 1, 1, 12, 12, 17, ...   
1                              [3, 3, 3, 2, 2, 3, 2]   
2  [10, 7, 8, 9, 9, 10, 10, 12, 15, 16, 19, 20, 1...   
3  [4, 6, 5, 2, 4, 4, 4, 4, 4, 4, 6, 13, 13, 12, ...   
4                                       [1, 1, 1, 1]   

                                                상영횟수  \
0  [2, 5, 8, 3, 1, 1, 1, 1, 2, 1, 1, 17, 17, 23, ...   
1                              [3, 3, 3, 2, 2, 3, 2]   
2  [14, 10, 12, 10, 15, 14, 13, 13, 17, 19, 26, 2...   
3  [4, 6, 7, 2, 5, 5, 4, 4, 6, 7, 10, 15, 15, 16,...   
4                                       [1, 1, 1, 1]   

                                                 조회일         개봉일 대표국적  ...  \
0  [2020-10-28 00:00:00, 2020-08-30 00:00:00, 202...  2020-06-24   한국  ...   
1  [2022-06-07 00:00:00, 2022-06-06 00:00:00, 202...  2022-06-01   한국  ...   
2  [2025-01-21 00:00:00, 2025-01-20 00:00:00, 202...  2025-01-08   한국  ...   
3  [2021-11-24 00:00:00, 2021-11-19 00:00:00, 202...  2021-09-29   미국  ...   
4  [2022-07-25 00:00:00, 2022-07-19 00:00:00, 202...        None   한국  ...   

        감독  매출액(점유율)      누적매출액  누적관객수 중복횟수  순위  매출액증감율(전일대비)  관객수증감율(전일대비)  \
0      조일형  0.218345     994000    142   58 NaN           NaN           NaN   
1      이상식  0.000143     274700     21    7 NaN           NaN           NaN   
2      오윤동  0.018133   34950000   1398   15 NaN           NaN           NaN   
3  캐리 후쿠나가  0.170660  220046880  28167   53 NaN           NaN           NaN   
4      한제이  0.000000       4000      1    4 NaN           NaN           NaN   

   제작사   배우  
0  NaN  NaN  
1  NaN  NaN  
2  NaN  NaN  
3  NaN  NaN  
4  NaN  NaN  

[5 rows x 24 columns]

In [12]:
result_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2517 entries, 0 to 2516
Data columns (total 24 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   영화명           2517 non-null   object 
 1   매출액증감         2517 non-null   object 
 2   관객수증감(전일대비)   2517 non-null   object 
 3   매출액           2517 non-null   object 
 4   관객수           2517 non-null   object 
 5   스크린수          2517 non-null   object 
 6   상영횟수          2517 non-null   object 
 7   조회일           2517 non-null   object 
 8   개봉일           1741 non-null   object 
 9   대표국적          2498 non-null   object 
 10  국적            2498 non-null   object 
 11  배급사           1840 non-null   object 
 12  등급            1801 non-null   object 
 13  장르            2491 non-null   object 
 14  감독            2427 non-null   object 
 15  매출액(점유율)      2517 non-null   float64
 16  누적매출액         2517 non-null   int64  
 17  누적관객수         2517 non-null   int64  
 18  중복횟수          2517 non-null 

In [13]:
result_df.isnull().sum()

영화명                0
매출액증감              0
관객수증감(전일대비)        0
매출액                0
관객수                0
스크린수               0
상영횟수               0
조회일                0
개봉일              776
대표국적              19
국적                19
배급사              677
등급               716
장르                26
감독                90
매출액(점유율)           0
누적매출액              0
누적관객수              0
중복횟수               0
순위              2517
매출액증감율(전일대비)    2517
관객수증감율(전일대비)    2517
제작사             2517
배우              2517
dtype: int64

### 앞의 형식으로 5년치 데이터 전체 전처리하여 csv로 저장했습니다.
### 다만 n번 이하로 나온 데이터에 대한 처리는 진행하지 않아서 총 영화 개수는 6650개 입니다.  

In [3]:
import pandas as pd

def merge_duplicate_movies_include_all(df):
    movie_col = '영화명'
    count_cols = ['매출액증감', '관객수증감(전일대비)']
    keep_cols = ['개봉일', '대표국적', '국적', '배급사', '등급', '장르', '감독']
    list_cols = ['매출액', '관객수', '스크린수', '상영횟수', '조회일']
    mean_cols = ['매출액(점유율)']
    last_cols = ['누적매출액', '누적관객수']

    count_cols = [col for col in count_cols if col in df.columns]
    keep_cols = [col for col in keep_cols if col in df.columns]
    list_cols = [col for col in list_cols if col in df.columns]
    mean_cols = [col for col in mean_cols if col in df.columns]
    last_cols = [col for col in last_cols if col in df.columns]

    if '조회일' in df.columns:
        df['조회일'] = pd.to_datetime(df['조회일'], errors='coerce')

    counts = df[movie_col].value_counts()
    dup_titles = counts[counts > 1].index

    df_dup = df[df[movie_col].isin(dup_titles)].sort_values(by=[movie_col, '조회일'], ascending=[True, False])

    agg_dict = {col: lambda x: list(x) for col in count_cols + list_cols}
    agg_dict.update({col: 'first' for col in keep_cols})
    agg_dict.update({col: 'mean' for col in mean_cols})
    agg_dict.update({col: 'last' for col in last_cols})

    df_merged = df_dup.groupby(movie_col).agg(agg_dict).reset_index()
    df_merged['중복횟수'] = df_dup[movie_col].value_counts()[df_merged[movie_col]].values

    df_unique = df[~df[movie_col].isin(dup_titles)].copy()
    df_unique['중복횟수'] = 1

    df_final = pd.concat([df_merged, df_unique], ignore_index=True)

    return df_final

# 🔽 CSV 파일로 저장하는 부분 추가
if __name__ == "__main__":
    # 예시: CSV 파일 불러오기
    df = pd.read_csv(r"C:\Users\김영우\기계학습_프로젝트\data\filled\filled_merged_all.csv")  # 🔁 입력 파일 경로 수정
    df_merged = merge_duplicate_movies_include_all(df)

    # 저장
    df_merged.to_csv("preprocessing_all.csv", index=False, encoding='utf-8-sig')
    print("✅ CSV 파일이 저장되었습니다: preprocessing_all.csv")


✅ CSV 파일이 저장되었습니다: preprocessing_all.csv
